File: 2121-find-if-path-exists-in-graph/find-if-path-exists-in-graph.cpp
class Solution {
public:
    bool bfs(vector<vector<int>>&adj,int n, int source, int des){
        vector<int>vis(n,0);
        queue<int>q;
        vector<int>pre(n,-1);
        q.push(source);
        vis[source]=1;
        while(!q.empty()){
            int node=q.front();q.pop();
            for(auto x:adj[node]){
                if(!vis[x]){
                    vis[x]=1;
                    pre[x]=node;
                    q.push(x);
                }
            }
        }
        if(vis[des]){
            return true;
        }
        return false;
    }
    bool validPath(int n, vector<vector<int>>& edges, int source, int des) {
        vector<vector<int>>adj(n);
        for(auto x:edges){
            adj[x[0]].push_back(x[1]);
            adj[x[1]].push_back(x[0]);
        }
        return bfs(adj,n,source,des);
    }
};

File: 2121-find-if-path-exists-in-graph/find-if-path-exists-in-graph.java
class Solution {
    public void solve(int node,ArrayList<ArrayList<Integer>> adj,int[] vis){
        vis[node]=1;
        for(int nbr: adj.get(node)){
            if(vis[nbr]==0){
                solve(nbr,adj,vis);
            }
        }
    }
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
        for(int i=0;i<n;i++){
            adj.add(new ArrayList<>());
        }
        for(int i=0;i<edges.length;i++){
            int[] edge = edges[i];
            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]);
        }
        int [] vis = new int[n];
        solve(source,adj,vis);
        if(vis[destination]==1){
            return true;
        }
        return false;
    }
}